### ISSUE TITLE
Implement missing endpoints

### ISSUE THREAD

**Issue by benediktwerner** [lichess-org/berserk#6]
These are the API endpoints that are currently still missing from berserk (generated by the `check-endpoints.py` script):

Feel free to choose any one and implement it (see https://github.com/lichess-org/berserk/blob/master/berserk/clients.py). For more info on the endpoint details, see <https://lichess.org/api>. Also see the [Get Started contributing](https://lichess-org.github.io/berserk/contributing.html#get-started) section.

- [ ] /api/bot/online
- [ ] /api/broadcast
- [ ] /api/broadcast/round/{broadcastRoundId}.pgn
- [ ] /api/broadcast/{broadcastTournamentId}.pgn
- [ ] /api/bulk-pairing
- [ ] /api/bulk-pairing/{id}
- [ ] /api/bulk-pairing/{id}/start-clocks
- [ ] /api/challenge/{challengeId}/cancel
- [ ] /api/challenge/{gameId}/start-clocks
- [ ] /api/cloud-eval
- [ ] /api/external-engine
- [ ] /api/external-engine/{id}
- [ ] /api/games/export/_ids
- [ ] /api/player
- [ ] /api/player/autocomplete
- [ ] /api/player/top/{nb}/{perfType}
- [ ] /api/puzzle/activity
- [ ] /api/puzzle/daily
- [ ] /api/puzzle/dashboard/{days}
- [ ] /api/puzzle/{id}
- [ ] /api/racer
- [ ] /api/round/{gameId}/add-time/{seconds}
- [ ] /api/storm/dashboard/{username}
- [ ] /api/stream/broadcast/round/{broadcastRoundId}.pgn
- [ ] /api/streamer/live
- [ ] /api/study/{studyId}.pgn
- [ ] /api/study/{studyId}/{chapterId}.pgn
- [ ] /api/swiss/{id}/edit
- [ ] /api/swiss/{id}/join
- [ ] /api/swiss/{id}/results
- [ ] /api/swiss/{id}/schedule-next-round
- [ ] /api/swiss/{id}/terminate
- [ ] /api/swiss/{id}/withdraw
- [ ] /api/team/all
- [ ] /api/team/of/{username}
- [ ] /api/team/search
- [ ] /api/team/{teamId}/request/{userId}/accept
- [ ] /api/team/{teamId}/request/{userId}/decline
- [ ] /api/team/{teamId}/requests
- [ ] /api/token/admin-challenge
- [ ] /api/tournament/team-battle/{id}
- [ ] /api/tournament/{id}/join
- [ ] /api/tournament/{id}/teams
- [ ] /api/tournament/{id}/terminate
- [ ] /api/tournament/{id}/withdraw
- [ ] /broadcast/{broadcastTournamentId}/edit
- [ ] /study/by/{username}/export.pgn
- [ ] /swiss/{id}.trf
- [ ] /team/{teamId}/pm-all
- [ ] https://engine.lichess.ovh/api/external-engine/work
- [ ] https://engine.lichess.ovh/api/external-engine/work/{id}
- [ ] https://engine.lichess.ovh/api/external-engine/{id}/analyse
- [ ] https://explorer.lichess.ovh/lichess/history
- [ ] https://explorer.lichess.ovh/master/pgn/{gameId}
- [ ] https://explorer.lichess.ovh/masters
- [ ] https://tablebase.lichess.ovh/antichess
- [ ] https://tablebase.lichess.ovh/atomic

**Comment by Virinas-code** [comment]
I might start working on this issue, thanks for the list!

**Comment by AKBuggy** [comment]
@Virinas-code May I work with you on this. I want to learn how you are implementing.

**Comment by Virinas-code** [comment]
Literrally copy pasting code and changing endpoints and docs...

**Comment by Virinas-code** [comment]
(I'm testing it after. It isn't that hard to understand, you just need a bit of python knowledge.)

**Comment by thegrapesoda** [comment]
@Virinas-code - do you have plans to implement /api/broadcast?  If not, I have an implementation ready to go.

**Comment by Virinas-code** [comment]
I'll get back to berserk dev soon. What should I start with?

**Comment by benediktwerner** [comment]
I'd say just pick any endpoint or group of endpoints that sound useful or interesting.

**Comment by Virinas-code** [comment]
Ok, gonna fork it again and work on tablebase I think.

**Comment by Virinas-code** [comment]
I confirm I'm working on it right now.

**Comment by zen1405** [comment]
Hi,

I opened a PR #30 addressing some of these.

**Comment by shaikhmudassir** [comment]
Hi @benediktwerner 
I am starting to work on this checklist. Have any changes been made to it? Since it hasn't been updated in three months. Also, the above [link for clients.py](https://github.com/lichess-org/berserk/blob/master/berserk/clients.py) is broken.

**Comment by rpesche** [comment]
I implemented one opening explorer endpoints: https://github.com/lichess-org/berserk/pull/32

**Comment by Virinas-code** [comment]
#43 Adding bulk pairing endpoints

**Comment by Virinas-code** [comment]
What is the `/api/racer` endpoint? I can't find it in the docs.

**Comment by benediktwerner** [comment]
https://lichess.org/api#tag/Puzzles/operation/racerPost

**Comment by Anupya** [comment]
A majority of the still unchecked items (as of Nov 2nd 2023, 7:46 pm ET) have already been implemented and exist!

Still unimplemented:
- `/api/cloud-eval`
- `/api/round/{gameId}/add-time/{seconds}`
- `/api/token`
- `/broadcast/{broadcastTournamentId}/edit`
- `/swiss/{id}.trf`
- `https://engine.lichess.ovh/api/external-engine/work`
- `https://engine.lichess.ovh/api/external-engine/work/{id}`
- `https://engine.lichess.ovh/api/external-engine/{id}/analyse`
- `https://explorer.lichess.ovh/lichess/history`

**Comment by kraktus** [comment]
thanks, double checked and updated

**Comment by jacksonthall22** [comment]
FYI, seems like `client.analysis.get_cloud_evaluation()` is not implemented (though it says it is in the readme): `AttributeError: 'Client' object has no attribute 'analysis'`

**Comment by benediktwerner** [comment]
That wasn't released yet, I released v0.13.2 with the recent changes.

**Comment by rdev99** [comment]
Is this issue still open ?

If its open, how much api's are left

**Comment by kraktus** [comment]
> Is this issue still open ?
> 
> If its open, how much api's are left

Hey, yes the endpoints not ticked are left to grab .

**Comment by rdev99** [comment]
Thanks, I will work on it if anyone else is not working

**Comment by kraktus** [comment]
Cool, if you want to implement external-engine endpoints, make sure to check feedback given in #66 before

**Comment by rdev99** [comment]
Hey,

I think these api's are already completed in the code but not ticked here.
```
https://explorer.lichess.ovh/master/pgn/{gameId}
https://explorer.lichess.ovh/masters
```

Also what is ```to be confirmed``` in ```https://explorer.lichess.ovh/lichess/history```

**Comment by FriedrichtenHagen** [comment]
@kraktus I opened a pr that implements the advice given in #66. Could you take a look?

**Comment by ShuvraneelMitra** [comment]
Hi! Any endpoints left to implement?

**Comment by fitztrev** [comment]
> Hi! Any endpoints left to implement?

Yes, there have been some new endpoints added the past few months that might not be in this list yet. You can check the changelog https://lichess.org/changelog

**Comment by FriedrichtenHagen** [comment]
@kraktus I added some endpoints in #86. Will keep working on the open endpoints afterwards.

**Comment by Virinas-code** [comment]
Thinking about the `oauth` endpoints. Maybe we can add a full flow system but using only Python's standard library (`http.server` and `webbrowser`)? This way we don't pull more dependencies but we can still make authentication easier, since it's often painful to implement.

**Comment by FriedrichtenHagen** [comment]
@benediktwerner  @kraktus Could you take a look at my open PRs? I would like to keep adding endpoints but I want to make sure I am on the right track first.

**Comment by Shoaib-Programmer** [comment]
@benediktwerner @kraktus I've completed the /api/puzzle/next endpoint. Please have a look at #99

**Comment by stafot** [comment]
Why fide get player and search players are not listed in missing entpoints, are they already implemented? Or is ` check-endpoints.py` failing to get them? 
It looks like they aren't implemented yet and I am keen to try add them if no objections.

**Comment by stafot** [comment]
PR to add FIDE endpoints -> https://github.com/lichess-org/berserk/pull/100

**Comment by Python-David** [comment]
> [@benediktwerner](https://github.com/benediktwerner) [@kraktus](https://github.com/kraktus) Could you take a look at my open PRs? I would like to keep adding endpoints but I want to make sure I am on the right track first.

Hey Ben did you get a response to those PRs?

**Comment by krndiy** [comment]
Hello, I'd like to help implementing, is the list of missing endpoints still valid?

**Comment by 01eksa** [comment]
Add the api/tournament/{id} endpoint to the issue.
